taichi/backends/opengl/opengl_api.h

```diff
@@ -88,14 +94,16 @@ struct CompiledProgram {
   std::unordered_map<int, irpass::ExternalPtrAccess> ext_arr_access;
   std::vector<std::string> str_table;
   UsedFeature used;
-  mutable std::unordered_map<int, CompiledArrayData> arr_args;
+  std::unordered_map<int, ScalarArg> scalar_args;
+  mutable std::unordered_map<int, CompiledArrayArg> arr_args;
```

Comment:
nit: I feel like we can remove `mutable` now?

Reply:
hmmm not yet for now, currently this is coupled with the JIT runtime logic "if runtime size doesn't change for **external** (note this is not ndarray) array, don't allocate_memory again, just reuse". 
We can consider decoupling external array and ndarray in the future. :D